### PR TITLE
DPE-3392 support rollback for unsupported datadir

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

1. deploy charm with mysql v0
2. refresh charm mysql v1
3. mysql v1 starts and make datadir incompat with prior versions
4. mysql v1 then fail to rejoin the cluster, rendering the unit in blocked state
5. attempt to rollback to previous v0
6. on rollback, mysql v0 fails to start, rendering the unit unusable

## Solution

- on rollback (step 5), detect specific `unsupported downgrade` error
- reset datadir, reinstall mysql and re-add as if was a new instance
- proceed with rollback

### CODE REVIEW

File: lib/charms/data_platform_libs/v0/upgrade.py [hunk 1]
```diff
@@ -606,6 +606,21 @@ def upgrade_stack(self, stack: List[int]) -> None:
         self.peer_relation.data[self.charm.app].update({"upgrade-stack": json.dumps(stack)})
         self._upgrade_stack = stack
 
+    @property
+    def other_unit_states(self) -> list:
+        """Current upgrade state for other units.
```
Comment (paulomach) [inline lib/charms/data_platform_libs/v0/upgrade.py:611]: these are a probable fix for an issue, still to be confirmed. 
If so, we will have a PR on upgrade libs before this get merged

File: lib/charms/data_platform_libs/v0/upgrade.py [hunk 2]
```diff
@@ -930,7 +945,6 @@ def on_upgrade_changed(self, event: EventBase) -> None:
             # on k8s only the upgrading unit will receive the upgrade-charm event
             # and deferring will prevent the upgrade stack from being popped
             logger.debug("Cluster in recovery, deferring...")
-            event.defer()
```
Comment (paulomach) [inline lib/charms/data_platform_libs/v0/upgrade.py:933]: The defer seems to be unnecessary, and causing a concurrency on `upgrade_granted` calls

**Review by taurus-forever — APPROVED** [review]
This is really nice. Thank you!

**Review by shayancanonical — APPROVED** [review]

File: src/mysql_vm_helpers.py [hunk 3]
```diff
@@ -195,6 +195,23 @@ def install_and_configure_mysql_dependencies() -> None:
...
+        """
+        try:
+            logger.debug("Uninstalling MySQL")
+            subprocess.run(["snap", "remove", "charmed-mysql"], check=True)
+        except subprocess.CalledProcessError:
+            # uninstalls fail due to SNAP_DATA_DIR fails to umount
+            # try umount it, without check
+            subprocess.run(["umount", CHARMED_MYSQL_COMMON_DIRECTORY])
+            logger.exception("Failed to uninstall MySQL")
+            raise MySQLInstallError
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/mysql_vm_helpers.py]: good point. It was also with the wrong exception
ffa2753

File: src/upgrade.py [hunk 4]
```diff
@@ -190,8 +190,24 @@ def _on_upgrade_granted(self, event: UpgradeGrantedEvent) -> None:
...
+        except MySQLStartMySQLDError:
+            # failed to start - check for a unsupported downgrade
+            if not self._check_server_unsupported_downgrade():
+                logger.error("Failed to start MySQL server after snap refresh")
+                self.set_unit_failed()
+                return
+
+            # on incompatible downgrade, a workload reset is required,
+            # but only if there's more then one unit, so SST can take place
+            if self.charm.app.planned_units() == 1:
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/upgrade.py:202]: we are using it extensively, but I see your point on ensuring that there are valid instances to recover from.

I'll add an improvement task for that. This is also valid in k8s

[the remark this answers is not in the document]
Comment (paulomach) [reply src/upgrade.py:202]: added on #412

**Pull request merged**